File: string/c++/delete_characters_from_string.cpp
/**
 Given a char array A, remove all chars that appears in char array B.
 */

#include <iostream>
using namespace std;

char* deleteChars(char* src, const char* toBeDelete) {
  if (src == NULL) return NULL;
  if (toBeDelete == NULL) return src;
  bool hashTable[256];
  memset(hashTable, false, 256);
  const char* p = toBeDelete;
  while (*p != '\0') {
    hashTable[*p] = true;
    p++;
  }
  int start = 0;
  int end = 0;
  while (src[start] != '\0') {
    while (hashTable[src[end]]) end++;
    src[start++] = src[end++];
  }
  return src;
}

int main() {
  char src[] = "I am a good man.";
  cout << "before: " << src << endl;
  cout << "after: " << deleteChars(src, "mdg") << endl;
  return 0;
}

File: string/java/DeleteCharactersFromString.java
/**
 Given a char array A, remove all chars that appears in char array B.
 */

import java.util.HashMap;

public class DeleteCharactersFromString {

  public static void main(String[] args) {
    System.out.println(deleteChars("abcdefgh", "cd"));
  }

  /*
    test cases -> expected output
    null, null -> null
    empty array, null -> empty array
    "abc", null -> "abc"
    "simple test", "st" -> "imple e"
   */
  public static String deleteChars(String src, String toBeDelete) {
    if (null == src) return null;
    if (toBeDelete == null) return src;
    // convert toBeDelete to hash map
    HashMap<Character, Byte> toBeDeleteHashMap = new HashMap<Character, Byte>();
    for (int i = 0; i < toBeDelete.length(); i++) toBeDeleteHashMap.put(toBeDelete.charAt(i), null);
    StringBuilder sb = new StringBuilder();
    // iterate through arr and delete char in toBeDeleteHashMap
    for (int i = 0; i < src.length(); i++) {
      char c = src.charAt(i);
      if (!toBeDeleteHashMap.containsKey(c)) sb.append(c);
    }
    return sb.toString();
  }
}
